plugins/module_utils/foreman_helper.py

```diff
@@ -590,10 +591,238 @@ def exit_json(self, changed=False, **kwargs):
         super(ForemanAnsibleModule, self).exit_json(**kwargs)
 
 
-class ForemanEntityAnsibleModule(ForemanAnsibleModule):
+class ResolverEntityMixin(object):
+    """ This mixin add automatic entities resolution based on provided attributes/ sub entities options.
+        It add following options to entity_spec 'entity' and 'entity_list' types:
+        * search_by (str): Field used to search the sub entity. Defaults to 'name'.
+        * search_operator (str): Operator used to search the su entity. Defaults to '='.
+        * resource_type (str): Resource type used to build API resource PATH. Defaults to pluralized entity key.
+        It add following attributes:
+        * entity_search (str): Defautls to None. If provided, the base entity resolver will use this search query instead to try to build it.
+        * entity_key (str): field used to search current entity. Defaults to 'name'. If entityèspec contains 'parent', defaults to 'title'.
+        * entity_scope (str): Type of the entity used to build search scope. Defaults to None.
+        * entity_resolve (boolean): Set it to False to disable base entity resolution.
+        * entity_opts (dict): Dict of options for base entity. Same options can be provided for subentities described in entity_spec.
+    """
+    def __init__(self, argument_spec=None, **kwargs):
+        self.original_entity_spec = kwargs.pop('entity_spec', {})
+        self.sub_entities = {key: value for key, value in self.original_entity_spec.items()
+                             if value.get('resolve', True) and 'type' in value and value['type'] in ['entity', 'entity_list']}
+        self.entity_scope = kwargs.pop('entity_scope', None)
+        self.entity_key = kwargs.pop('entity_key', 'name')
+        self.entity_resolve = kwargs.pop('entity_resolve', True)
+        self.entity_opts = kwargs.pop('entity_opts', {})
+        if 'thin' not in self.entity_opts:
+            self.entity_opts['thin'] = False
+        if 'failsafe' not in self.entity_opts:
+            self.entity_opts['failsafe'] = True
+        if 'parent' in self.sub_entities:
+            self.entity_opts['search_by'] = 'title'
+        elif 'search_by' not in self.entity_opts:
+            self.entity_opts['search_by'] = self.entity_key
+
+        # Convert current class name from CamelCase to snake_case
+        class_name = re.sub(r'(?<=[a-z])[A-Z]|[A-Z](?=[^A-Z])', r'_\g<0>', self.__class__.__name__).lower().strip('_')
+        # Get entity name from snake case class name
+        self.entity_name = '_'.join(class_name.split('_')[1:-1])
```

Comment:
I wonder if this is too much magic and we should just pass `entity_name` to `__init__`?

Reply:
As you want, the spirit was "convention over configuration", and it feels redundant with the class name for me.

Reply:
I wouldn't block on this, the magic way is fine too :)

Reply:
I'd at least have a helper `get_entity_name_from_class` that you can individually test. You can also allow passing an `entity_name` to `__init__` but fall back on the automatic way.

Reply:
Moved the 'magic' to method and added ability to pass it to `__init__`